[MRG] Adding variable force_alpha to classes in naive_bayes.py

hongshaoyang opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes #10772
References PR #9131
References PR #10775
Taking over stalled PR #16747


#### What does this implement/fix? Explain your changes.

This PR takes over stalled PR #16747.

From the description of #16747: "This PR adds a new variable alphaCorrection in classes in naive_bayes.py, which is set to True by default and if set to False, then for alpha=0 (or greater, but still smaller than _ALPHA_MIN) alpha is not being rounded up to _ALPHA_MIN."

I merged master and fixed conflicts, no other changes.

#### Any other comments?

https://github.com/scikit-learn/scikit-learn/pull/16747#discussion_r432946153 should be resolved.
<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

arka204 committed:
Adding variable alphaCorrection to classes in naive_bayes.py.

arka204 committed:
Splitting few lines of code.

arka204 committed:
Merge pull request #1 from scikit-learn/master

Merging changes from the main repository

arka204 committed:
Merge pull request #2 from scikit-learn/master

Merging changes from the main repository

arka204 committed:
Fixing problems and adding tests.

arka204 committed:
Updating naive_bayes.py.

arka204 committed:
Merge pull request #3 from scikit-learn/master

Merging changes from the main repository

arka204 committed:
Merge branch 'alpha-close-or-equal-0-update' into alpha-1

arka204 committed:
Merge pull request #5 from arka204/alpha-1

Update branch

arka204 committed:
Merge pull request #6 from arka204/alpha-close-or-equal-0-update

Resolving conflicts

arka204 committed:
Checkig warnings in tests.

arka204 committed:
Merge branch 'alpha-close-or-equal-0-update' of https://github.com/arka204/scikit-learn into alpha-close-or-equal-0-update

arka204 committed:
Merge pull request #8 from arka204/alpha-close-or-equal-0-update

Alpha close or equal 0 update

arka204 committed:
Update v0.24.rst

arka204 committed:
Merge pull request #10 from scikit-learn/master

Update master

arka204 committed:
Merge branch 'Proposition-for-BernoulliNB-and-MultinomialNB-when-alpha-is-close-or-equal-0' into master-copy

arka204 committed:
Merge pull request #11 from arka204/master-copy

Update branch

Commit:
Merge remote-tracking branch 'upstream/master' into Proposition-for-BernoulliNB-and-MultinomialNB-when-alpha-is-close-or-equal-0

# Conflicts:
#	doc/whats_new/v0.24.rst
#	sklearn/naive_bayes.py

Commit:
Fix merge

hongshaoyang commented:
Ping @cmarmo @jnothman @amueller

hongshaoyang commented:
Gentle ping @cmarmo @jnothman @amueller

cmarmo commented:
Hi @hongshaoyang , sorry your pull request didn't make it in 0.24. Do you mind moving your what's new entry from v0.24.rst to v1.0.rst? Thanks!

hongshaoyang committed:
Merge remote-tracking branch 'upstream/master' into 10772-force-alpha

hongshaoyang committed:
Move whatsnew

hongshaoyang committed:
Merge remote-tracking branch 'upstream/master' into 10772-force-alpha

# Conflicts:
#	doc/whats_new/v1.0.rst

hongshaoyang committed:
Merge remote-tracking branch 'upstream/main' into 10772-force-alpha

jjerphan commented on sklearn/naive_bayes.py:
```diff
@@ -822,7 +834,13 @@ class ComplementNB(_BaseDiscreteNB):
+        Additive (Laplace/Lidstone) smoothing parameter
+        (set alpha=0 and force_alpha=True, for no smoothing).
+
+    force_alpha : bool, default=False
+        If false and alpha is too close to 0, it will set alpha to _ALPHA_MIN.
+        If true, warn user about potential numeric errors
+        and proceed with alpha unchanged.
```
Just a small correction.

```suggestion
    force_alpha : bool, default=False
        If False and alpha is too close to 0, it will set alpha to `_ALPHA_MIN`.
        If True, warn user about potential numeric errors
        and proceed with alpha unchanged.

        .. versionadded:: 0.24
```

jjerphan commented on sklearn/naive_bayes.py:
```diff
@@ -951,7 +970,12 @@ class BernoulliNB(_BaseDiscreteNB):
-        (0 for no smoothing).
+        (set alpha=0 and force_alpha=True, for no smoothing).
+
+    force_alpha : bool, default=False
+        If false and alpha is too close to 0, it will set alpha to _ALPHA_MIN.
+        If true, warn user about potential numeric errors
+        and proceed with alpha unchanged.
```
Ditto.
```suggestion
    force_alpha : bool, default=False
        If False and alpha is too close to 0, it will set alpha to `_ALPHA_MIN`.
        If True, warn user about potential numeric errors
        and proceed with alpha unchanged.

        .. versionadded:: 0.24
```

jjerphan commented on sklearn/naive_bayes.py:
```diff
@@ -703,7 +708,12 @@ class MultinomialNB(_BaseDiscreteNB):
-        (0 for no smoothing).
+        (set alpha=0 and force_alpha=True, for no smoothing).
+
+    force_alpha : bool, default=False
+        If false and alpha is too close to 0, it will set alpha to _ALPHA_MIN.
+        If true, warn user about potential numeric errors
+        and proceed with alpha unchanged.
```
```suggestion
    force_alpha : bool, default=False
        If False and alpha is too close to 0, it will set alpha to _ALPHA_MIN.
        If True, warn user about potential numeric errors
        and proceed with alpha unchanged.
        
        .. versionadded:: 0.24
```

jjerphan commented on sklearn/naive_bayes.py:
```diff
@@ -1086,7 +1111,12 @@ class CategoricalNB(_BaseDiscreteNB):
+        (set alpha=0 and force_alpha=True, for no smoothing).
+
+    force_alpha : bool, default=False
+        If false and alpha is too close to 0, it will set alpha to _ALPHA_MIN.
+        If true, warn user about potential numeric errors
```
Ditto.
```suggestion
    force_alpha : bool, default=False
        If False and alpha is too close to 0, it will set alpha to `_ALPHA_MIN`.
        If True, warn user about potential numeric errors
        and proceed with alpha unchanged.

        .. versionadded:: 0.24
```

jjerphan reviewed (changes requested):
Thanks @hongshaoyang to supersede this.

LGTM after some small docstrings adaptation.

jjerphan commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -868,6 +868,25 @@ def test_alpha():
+    msg2 = ('alpha too small will result in numeric errors, '
+            'setting alpha = %.1e' % _ALPHA_MIN)
+    b = BernoulliNB(alpha=0, force_alpha=True)
+    with pytest.warns(UserWarning, match=msg1):
+        assert b._check_alpha() == 0
+    b = BernoulliNB(alpha=0, force_alpha=False)
+    with pytest.warns(UserWarning, match=msg2):
+        assert b._check_alpha() == _ALPHA_MIN
+    b = BernoulliNB(alpha=0)
+    with pytest.warns(UserWarning, match=msg2):
+        assert b._check_alpha() == _ALPHA_MIN
```
You can use `assert_warns_message` to assert that warnings are thrown in those cases. 🙂

hongshaoyang replied:
i'm following #19414 to not use assert_warns_message - this will allow me to use regex for the warning message

jjerphan replied:
Indeed, this is a better practice :+1:

hongshaoyang committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jjerphan commented on sklearn/naive_bayes.py:
```diff
@@ -1114,8 +1120,12 @@ class CategoricalNB(_BaseDiscreteNB):
     force_alpha : bool, default=False
-        If false and alpha is too close to 0, it will set alpha to _ALPHA_MIN.
-        If true, warn user about potential numeric errors
+    force_alpha : bool, default=False
```
This line has been duplicated here.

hongshaoyang committed:
Remove extra line

hongshaoyang committed:
Flake8

jjerphan commented on sklearn/naive_bayes.py:
```diff
@@ -703,7 +708,14 @@ class MultinomialNB(_BaseDiscreteNB):
-        (0 for no smoothing).
+        (set alpha=0 and force_alpha=True, for no smoothing).
+
+    force_alpha : bool, default=False
+        If False and alpha is too close to 0, it will set alpha to _ALPHA_MIN.
+        If True, warn user about potential numeric errors
+        and proceed with alpha unchanged.
```
For consistency:
```suggestion
    force_alpha : bool, default=False
        If False and alpha is too close to 0, it will set alpha to
        `_ALPHA_MIN`. If True, warn user about potential numeric errors
        and proceed with alpha unchanged.
```

hongshaoyang committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

hongshaoyang committed:
Merge remote-tracking branch 'upstream/main' into 10772-force-alpha

# Conflicts:
#	sklearn/naive_bayes.py

hongshaoyang committed:
Fix merge

hongshaoyang committed:
use assert_warns_message

jjerphan commented on sklearn/naive_bayes.py:
```diff
@@ -526,9 +526,14 @@ def _check_alpha(self):
-                          'setting alpha = %.1e' % _ALPHA_MIN)
-            return np.maximum(self.alpha, _ALPHA_MIN)
+            if self.force_alpha:
+                warnings.warn('alpha too small will result in numeric errors, '
+                              'force_alpha was set to True, '
+                              'proceeding without changing alpha.')
```
I think that we can report the value of `alpha` in the warning, making it easier users to inspect its value in the case of numeric errors. Also here is a small rewording proposals (not using f-string for the consistency):
```suggestion
                warnings.warn('alpha too small will result in numeric errors. '
                              'Proceeding with alpha = %.1e, as '
                              'force_alpha was set to True.' % alpha)
```

hongshaoyang replied:
Updated.

jjerphan reviewed (changes requested):
I am catching up with this PR.
I just have one last suggestion prior to the merge.

hongshaoyang committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

hongshaoyang committed:
Merge remote-tracking branch 'upstream/main' into 10772-force-alpha

hongshaoyang committed:
Fix wrong variable name

hongshaoyang committed:
Fix test to use "with pytest.warns" instead of assert_warns_message

jjerphan commented:
Small tip: I would suggest to both make modifications and run tests locally prior to pushing new commits.

This would allow you better take suggestions into account, group suggestions (if relevant) and assert if you've missed something (as suggestions sometimes necessitates other changes other than in their patches) without going back and forth with the CI logs. :slightly_smiling_face:

hongshaoyang commented:
Thank you!  I was eager to commit the suggestions without taking tests into account. Your suggestion is helpful!

thomasjpfan committed:
MAINT Adds target_version to black config (#20293)

jjerphan commented:
Hi @hongshaoyang,

In the meantime, black code formatting was set on `main`, causing conflict on this branch.

Instructions given in https://github.com/scikit-learn/scikit-learn/issues/20301#issue-925041705 might help you with resolving them.

hongshaoyang committed:
Merge commit '0e7761cdc4f244adb4803f1a97f0a9fe4b365a99' into 10772-force-alpha

hongshaoyang committed:
Black formatting

Following
https://github.com/scikit-learn/scikit-learn/issues/20301#issue-925041705

hongshaoyang committed:
Merge remote-tracking branch 'upstream/main' into 10772-force-alpha

# Conflicts:
#	sklearn/naive_bayes.py

jjerphan commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -897,6 +897,29 @@ def test_alpha():
 
 
+def test_check_alpha():
+    # Test force_alpha if alpha < _ALPHA_MIN
```
```suggestion
    # Non-regression test for:
    # https://github.com/scikit-learn/scikit-learn/issues/10772
    # Test force_alpha if alpha < _ALPHA_MIN
```

jjerphan reviewed (approved):
LGTM @hongshaoyang. :+1: 

I just have one last suggestion.

hongshaoyang committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

cmarmo commented:
Hi @hongshaoyang, and thanks for your patience. If you are still motivated, this PR could be a nice fix for version 1.1.
Do you mind fixing conflicts and updating the version number to 1.1? Then let's see if  perhaps @TomDLT might be available for a quick review? Thanks!

TomDLT commented on sklearn/naive_bayes.py:
```diff
@@ -537,11 +537,18 @@ def _check_alpha(self):
+                )
+            else:
+                warnings.warn(
+                    "alpha too small will result in numeric errors, "
```
I would add something like "use force_alpha=True to keep alpha unchanged.".

TomDLT reviewed (commented):
Some comments. 

Main question: Do we really need to warn when the user explicitly set `force_alpha=True` ?

TomDLT commented on doc/whats_new/v1.0.rst:
```diff
@@ -462,6 +462,11 @@ Changelog
 :mod:`sklearn.naive_bayes`
 ..........................
+- |Fix| A new parameter `force_alpha` was added to :class:`BernoulliNB` and
+  class:`MultinomialNB`, allowing user to set parameter alpha to a very
```
Also need to mention `ComplementNB` and `CategoricalNB`.

Micky774 commented:
Picked up in PR #22269

cmarmo commented:
Closing this as superseded by #22269 .
